### ISSUE TITLE
rwanalysis error messages

### ISSUE THREAD

**Issue by SocialHulkHogan** [WheatonCS/Lexos#408]
So im trying to put in an error message so the page doesn't crash on this page. To do that, I want to check to see if certain variables are checked before the graph is drawn. the timing of the makeRWAGraph() makes no sense, so I was trying to change the code so the checks are done before the graph is made. I was trying to make it so the graph wasn't drawn unless certain variables are filled out, but no luck.

**Comment by scottkleinman** [comment]
That helps. I'll try to figure this out.

**Comment by SocialHulkHogan** [comment]
the get graph button doesn't work anymore when you give it the id get graph.

**Comment by SocialHulkHogan** [comment]
I just pushed. Please pull. If you don't see two get graph buttons, you didn't pull successfully. Please make sure you have pulled successfully. We need to be working with the same code

**Comment by scottkleinman** [comment]
I can see the buttons.

**Comment by SocialHulkHogan** [comment]
Ok good. 

<script>
$(document).ready(function() {
   $("#getgraph").click(function(e){
    e.preventDefault();
    // // This is just a sample variable to test against. Change it to 0 to generate an error modal.
    numActiveDocs = 1;
    if (numActiveDocs < 1) {
      msg = 'You have no active documents. Please activate at least one document using the <a href="./manage">Manage</a> tool or <a href="./upload">upload</a> a new document.';
      $('#error-modal-message').html(msg);
      $('#error-modal').modal();
      return false;
    }
    else {
      $("form").attr("method", "post")
      $("form").submit();
    }
  });
});
</script>

this is the code you had me add to the bottom it breaks button number 2 (more left)

**Comment by SocialHulkHogan** [comment]
$(document).ready(function() {
   $("#getgraph").click(function(e){
    e.preventDefault();
    // // This is just a sample variable to test against. Change it to 0 to generate an error modal.
    numActiveDocs = 1;
    if (numActiveDocs < 1) {
      msg = 'You have no active documents. Please activate at least one document using the <a href="./manage">Manage</a> tool or <a href="./upload">upload</a> a new document.';
      $('#error-modal-message').html(msg);
      $('#error-modal').modal();
      return false;
    }
    else {
      $("form").attr("method", "post")
      $("form").submit();
    }
  });
});

**Comment by SocialHulkHogan** [comment]
I understand this a lot more now. the issue really seems to boil down to giving the get graph button a click event stops the post request that is triggered by the input type= "submit"

**Comment by SocialHulkHogan** [comment]
so without that code from two messages ago, if you make the makeRWAGraph() function in an onclick based on the id of the button, the page is rerendered because a submit happens and thats why the graph only flashes and is destroyed. I'm not sure how to avoid this.

**Comment by scottkleinman** [comment]
Damn! I answered this but apparently failed to press the Comment button. Hopefully you saw what I put in the code when I pushed a fix. I'm going to pull again and see where you are.

**Comment by mleblanc321** [comment]
(steven has left for the evening; i'm not sure if he pushed any changes he was trying)

**Comment by scottkleinman** [comment]
I can see the phenomenon he describes, so I'll try to figure out what's causing it.

**Comment by mleblanc321** [comment]
steven just pushed a change (actually, on my box) with an IF-ELSE at the end of the POST in lexos.py for def rollingwindow() ...  (the difference only being if we set rwadatagenerated = [True or False] )

now he thinks we can use this to properly display a modal message in the event that the button is clicked and the rollingWindowSize field is left empty

**Comment by scottkleinman** [comment]
Oh, I've also got a tentative fix to the problem and was just cleaning up code. I'll compare the two tonight. My solution was actually to forget about rwadatagenerated and simply test in the Javascript whether any data has been passed to the DOM (I check the length of the readLines array).

**Comment by SocialHulkHogan** [comment]
because if you try to check in the js you can't just then not call make graph because the page will still transform on an html level even if you stop the graph.

**Comment by SocialHulkHogan** [comment]
I'm going to eat then go back to work I think I can get this with another hour or two.

**Comment by SocialHulkHogan** [comment]
Finally! I finally have something resembling a solution! It isn't a modal, but the page does throw an alert with the text "boop" at the exact right time. Now I just have to make it so it doesn't re-render the page in this case but for the most part this currently functions perfectly.

**Comment by SocialHulkHogan** [comment]
we could even say in the error message "please wait for the page to reload after closing this" if stopping the reload is impossible

**Comment by SocialHulkHogan** [comment]
let me know if there is a better way, I'm just glad I got to a working solution at this point.

**Comment by SocialHulkHogan** [comment]
If you could add some of that form submit false code to this function

 $("#getgraph").click(function() {
    // Do some validation here... I often serialise the form values as a json string for reference
    // If the form passes...
//window.location="#graph-anchor";

if($("#rollingwindowsize").value==null)
{
	alert("boop");
}
});
in rwanalysis.js that'd be great.

**Comment by scottkleinman** [comment]
I just had a 90 minute drive, but I'll see what I can do before tomorrow.

**Comment by scottkleinman** [comment]
Well, it appears that the code I wrote before I left for Pasadena works perfectly. I'll pull later, compare it to yours, and then try to integrate the various error messages.

Oops! I spoke to soon. Chrome doesn't seem to be submitting the form values. I'll work on that too.

**Comment by scottkleinman** [comment]
Grrrr... Chrome does not accept jQuery `$("form").submit()`, but it is happy with pure Javascript: `document.forms[0].submit();`. But at least that means I can move on...

**Comment by SocialHulkHogan** [comment]
Can you give me a chance to make back ups of what I have atm before you push? It's at a good spot and I don't want to lose that in case the new solution turns out to be flawed somehow.

**Comment by SocialHulkHogan** [comment]
On a side note Scott this is wrapping up nicely on the front end. Thanks for all your help it's been a pleasure working with you

**Comment by scottkleinman** [comment]
That's OK. I wasn't planning to push anything before seeing your code. Yours seems to work, but I can't see what exactly you've changed. I'll take a second look tomorrow, since it's now late here. If you want to look at my code in the morning, I pushed files called `rwanalysis2.html` and `scripts_rwanalysis2.js`. They won't actually work unless you change the names back, but you can see what I've done. The validation is actually a bit weird. Invalid setting combinations trigger error messages _before_ you submit, so that's slightly different from the other tools.

Good work!

**Comment by SocialHulkHogan** [comment]
I don't think your push went through, I can't find the files you are referring to

**Comment by scottkleinman** [comment]
Uh, no idea what happened. I've pushed them now.

**Comment by SocialHulkHogan** [comment]
The current code actually works perfectly now. the alert is a modal, triggers perfectly and stops page re-render. e.preventDefault() is all that was needed. the html rendering problems were the main thing in the way of doing this and I figured that out last night. In filemanager or lexos.py or utilities (i forget) I made it so the page didn't render as if raw was true if that field was blank. After that, 

$("#getgraph").click(function(e) {
if($("#rollingwindowsize")[0].value=="")
	{	e.preventDefault();
	 msg = "Please fill out the 'Size of Rolling Window' and 'Search Pattern(s)' fields";
      $('#error-modal-message').html(msg);
      $('#error-modal').modal();
	}
	});

This simple function accomplishes our goal.
